shared/test/TestTileSource.cpp

```diff
@@ -0,0 +1,95 @@
+                           std::make_shared<WebMercatorTiled2dMapLayerConfig>(
+                               "mock", "{z}/{x}/{y}", Tiled2dMapZoomInfo(1.0, 0, 0, false, true, false, true), 0, 20),
+                           CoordinateConversionHelperInterface::independentInstance(), scheduler, 62, 1, "layer") {};
+
+    std::unordered_set<Tiled2dMapTileInfo> getCurrentTiles() {
+        std::unordered_set<Tiled2dMapTileInfo> tiles;
+        for (const auto &tile : currentTiles) {
+            tiles.insert(tile.first);
+        }
+        return tiles;
+    }
+
+    void notifyTilesUpdates() override {}
+
+  protected:
+    void cancelLoad(Tiled2dMapTileInfo tile, size_t loaderIndex) override {}
+
+    ::djinni::Future<std::shared_ptr<TestLoaderResult>> loadDataAsync(Tiled2dMapTileInfo tile, size_t loaderIndex) override {
+        auto promise = std::make_shared<::djinni::Promise<std::shared_ptr<TestLoaderResult>>>();
+        promise->setValue(std::make_shared<TestLoaderResult>());
+        return promise->getFuture();
+    }
+
+    bool hasExpensivePostLoadingTask() override { return false; }
+
+    int postLoadingTask(std::shared_ptr<TestLoaderResult> loadedData, Tiled2dMapTileInfo tile) override { return 12; }
+};
+
+TEST_CASE("VectorTileSource") {
+    auto scheduler = std::make_shared<TestScheduler>();
+    std::shared_ptr<TestTiled2dMapVectorSource> source = std::make_shared<TestTiled2dMapVectorSource>(scheduler);
+    source->mailbox = std::make_shared<Mailbox>(scheduler);
+
+    SECTION("basic rect") {
+        auto rect = RectCoord(Coord(3857, 409758, 6263475, 0), Coord(3857, 1390769, 5579492, 0));
+        source->onVisibleBoundsChanged(rect, 0, 4371851);
+        scheduler->drain();
+
+        auto tiles = source->getCurrentTiles();
+
+        // ignore the rect for assertions
+        RectCoord empty = rect;
+        std::vector<Tiled2dMapTileInfo> expectedTiles = {{empty, 0, 0, 0, 0, 559082264}, {empty, 32, 22, 0, 6, 8735660},
+                                                         {empty, 33, 21, 0, 6, 8735660}, {empty, 34, 23, 0, 6, 8735660},
+                                                         {empty, 32, 21, 0, 6, 8735660}, {empty, 33, 23, 0, 6, 8735660},
+                                                         {empty, 32, 23, 0, 6, 8735660}, {empty, 33, 22, 0, 6, 8735660},
+                                                         {empty, 34, 21, 0, 6, 8735660}, {empty, 34, 22, 0, 6, 8735660}};
+
+        REQUIRE(tiles.size() == 10);
+        for (const auto &expectedTile : expectedTiles) {
+            REQUIRE(std::find_if(tiles.begin(), tiles.end(), [&expectedTile](const auto &tile) {
+                        return tile.x == expectedTile.x && tile.y == expectedTile.y &&
+                               tile.zoomIdentifier == expectedTile.zoomIdentifier && tile.zoomLevel == expectedTile.zoomLevel;
+                    }) != tiles.end());
+        }
```

Comment:
I think this could use https://github.com/catchorg/Catch2/blob/devel/docs/matchers.md#generic-range-matchers?

```suggestion
        REQUIRE_THAT(tiles, Catch::Matchers::UnorderedRangeEquals(expectedTiles));
```

Requires `#include <catch2/matchers/catch_matchers_range_equals.hpp>`

Reply:
Btw. in failed asserts the tile info structs currently just print as `{?}`. To fix this, we can define stream out operator in `shared/public/Tiled2dMapTileInfo.h`:
```cpp
inline std::ostream& operator<<(std::ostream& os, Tiled2dMapTileInfo& tile)
{
    return os << tile.to_string();
}
```